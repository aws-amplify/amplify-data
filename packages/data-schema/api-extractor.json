{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
