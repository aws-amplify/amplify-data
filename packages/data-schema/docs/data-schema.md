<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
