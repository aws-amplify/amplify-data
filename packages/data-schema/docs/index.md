<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [@aws-amplify/data-schema](./data-schema.md) |  |

