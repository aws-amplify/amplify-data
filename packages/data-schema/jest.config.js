/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  testPathIgnorePatterns: [
    ".test-d.ts", // don't directly execute these files. They'll be tested via their `test.ts` counterparts
    "/node_modules/",
    "utils.ts",
    "/fixtures/",
    "/MappedTypes/",
  ],
  coverageThreshold: {
		global: {
			branches: 72,
			functions: 86,
			lines: 85,
			statements: 85,
		},
	},
};
