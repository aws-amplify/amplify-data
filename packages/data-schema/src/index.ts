import { schema } from './ModelSchema';
import { model } from './ModelType';
import {
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
} from './ModelField';
import { ref } from './RefType';
import { hasOne, hasMany, belongsTo, manyToMany } from './ModelRelationalField';
import { allow } from './Authorization';
import { customType } from './CustomType';
import { enumType } from './EnumType';
import { query, mutation, subscription } from './CustomOperation';
import { handler } from './Handler';

// We are re-exporting with this pattern to make the JSDoc happy.
// Otherwise the JSDoc wouldn't consistently show up as a customer types
// "a."
export {
  schema,
  model,
  ref,
  customType,
  enumType as enum,
  query,
  mutation,
  subscription,
  hasOne,
  hasMany,
  belongsTo,
  manyToMany,
  allow,
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
  handler,
};
