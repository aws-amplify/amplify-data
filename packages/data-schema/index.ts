import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
