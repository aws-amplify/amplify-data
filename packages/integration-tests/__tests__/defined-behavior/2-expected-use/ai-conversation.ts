import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  subOptionsAndHeaders,
  pause,
} from '../../utils';

describe('AI Conversation Routes', () => {
  // data/resource.ts
  const schema = a.schema({
    chatBot: a.conversation({
      aiModel: a.ai.model('Claude 3 Haiku'),
      systemPrompt: 'You are a helpful chatbot.',
    }),
  });
  type Schema = ClientSchema<typeof schema>;

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Conversations', () => {
    test('Create a conversation', async () => {
      // #region mocking
      const sampleConversation = {
        id: 'conversation-id',
        createdAt: '2023-06-01T12:00:00Z',
        updatedAt: '2023-08-02T12:00:00Z',
        metadata: {},
        name: 'Test Conversation',
      };
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            createConversation: sampleConversation,
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region api call
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // create conversation
      const { data: createdConversation, errors: createConversationErrors } =
        await client.conversations.chatBot.create({
          name: 'Test Conversation',
          metadata: { arbitrary: 'data' },
        });
      // #endregion api call

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(createConversationErrors).toBeUndefined();
      expect(createdConversation).toStrictEqual({
        createdAt: '2023-06-01T12:00:00Z',
        id: sampleConversation.id,
        listMessages: expect.any(Function),
        metadata: {},
        name: 'Test Conversation',
        onStreamEvent: expect.any(Function),
        sendMessage: expect.any(Function),
        updatedAt: '2023-08-02T12:00:00Z',
      });
      // #endregion assertions
    });

    test('Update a conversation', async () => {
      const sampleConversation = {
        id: 'conversation-id',
        createdAt: '2023-06-01T12:00:00Z',
        updatedAt: '2023-08-02T12:00:00Z',
        metadata: {},
        name: 'Test Conversation',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            updateConversation: sampleConversation,
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region api call
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // create conversation
      const { data: updatedConversation, errors: updateConversationErrors } =
        await client.conversations.chatBot.update({
          id: sampleConversation.id,
          name: 'updated conversation name',
          metadata: { arbitrary: 'data' },
        });
      // #endregion api call

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(updateConversationErrors).toBeUndefined();
      expect(updatedConversation).toStrictEqual({
        createdAt: '2023-06-01T12:00:00Z',
        id: sampleConversation.id,
        listMessages: expect.any(Function),
        metadata: {},
        name: 'Test Conversation',
        onStreamEvent: expect.any(Function),
        sendMessage: expect.any(Function),
        updatedAt: '2023-08-02T12:00:00Z',
      });
      // #endregion assertions
    });

    test('Update a conversation', async () => {
      const sampleConversation = {
        id: 'conversation-id',
        createdAt: '2023-06-01T12:00:00Z',
        updatedAt: '2023-08-02T12:00:00Z',
        metadata: {},
        name: 'Test Conversation',
      };

      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            updateConversation: sampleConversation,
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region api call
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // create conversation
      const { data: updatedConversation, errors: updateConversationErrors } =
        await client.conversations.chatBot.update({
          id: sampleConversation.id,
          name: 'updated conversation name',
          metadata: { arbitrary: 'data' },
        });
      // #endregion api call

      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(updateConversationErrors).toBeUndefined();
      expect(updatedConversation).toStrictEqual({
        createdAt: '2023-06-01T12:00:00Z',
        id: sampleConversation.id,
        listMessages: expect.any(Function),
        metadata: {},
        name: 'Test Conversation',
        onStreamEvent: expect.any(Function),
        sendMessage: expect.any(Function),
        updatedAt: '2023-08-02T12:00:00Z',
      });
      // #endregion assertions
    });

    test('Get a conversation', async () => {
      // #region mocking
      const sampleConversation = {
        id: 'conversation-id',
        createdAt: '2023-06-01T12:00:00Z',
        updatedAt: '2023-06-01T12:00:00Z',
        metadata: {},
        name: 'Test Conversation',
      };
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getConversation: sampleConversation,
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // get conversation
      const { data: conversation, errors: getConversationErrors } =
        await client.conversations.chatBot.get({ id: sampleConversation.id });
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(getConversationErrors).toBeUndefined();
      expect(conversation).toStrictEqual({
        createdAt: '2023-06-01T12:00:00Z',
        id: sampleConversation.id,
        listMessages: expect.any(Function),
        metadata: {},
        name: 'Test Conversation',
        onStreamEvent: expect.any(Function),
        sendMessage: expect.any(Function),
        updatedAt: '2023-06-01T12:00:00Z',
      });
      // #endregion assertions
    });

    test('List conversations', async () => {
      // #region mocking
      const sampleConversation1 = {
        id: 'conversation-id-1',
        createdAt: '2023-06-01T12:00:00Z',
        updatedAt: '2023-08-02T12:00:00Z',
        metadata: {},
        name: 'Test Conversation',
      };
      const sampleConversation2 = {
        id: 'conversation-id-2',
        createdAt: '2024-09-02T12:00:00Z',
        updatedAt: '2024-09-05T12:00:00Z',
        metadata: {},
        name: 'Test Conversation2',
      };
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            listConversations: {
              items: [sampleConversation1, sampleConversation2],
            },
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      // list conversations
      const { data: conversations, errors: listConversationsErrors } =
        await client.conversations.chatBot.list();
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(listConversationsErrors).toBeUndefined();
      expect(conversations).toStrictEqual([
        {
          createdAt: '2023-06-01T12:00:00Z',
          id: sampleConversation1.id,
          listMessages: expect.any(Function),
          metadata: {},
          name: 'Test Conversation',
          onStreamEvent: expect.any(Function),
          sendMessage: expect.any(Function),
          updatedAt: '2023-08-02T12:00:00Z',
        },
        {
          createdAt: '2024-09-02T12:00:00Z',
          id: sampleConversation2.id,
          listMessages: expect.any(Function),
          metadata: {},
          name: 'Test Conversation2',
          onStreamEvent: expect.any(Function),
          sendMessage: expect.any(Function),
          updatedAt: '2024-09-05T12:00:00Z',
        },
      ]);
      // #endregion assertions
    });

    test('Paginate conversations', async () => {
      // #region mocking
      const sampleConversation1 = {
        id: 'conversation-id-1',
        createdAt: '2023-06-01T12:00:00Z',
        updatedAt: '2023-08-02T12:00:00Z',
        metadata: {},
        name: 'Test Conversation',
      };
      const sampleConversation2 = {
        id: 'conversation-id-2',
        createdAt: '2024-09-02T12:00:00Z',
        updatedAt: '2024-09-05T12:00:00Z',
        metadata: {},
        name: 'Test Conversation2',
      };
      const sampleNextToken = 'next-token';
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            listConversations: {
              items: [sampleConversation1, sampleConversation2],
              nextToken: sampleNextToken,
            },
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      const {
        data: conversations,
        nextToken,
        errors: listConversationsErrors,
      } = await client.conversations.chatBot.list({
        limit: 100,
        nextToken: 'previous-next-token',
      });
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(listConversationsErrors).toBeUndefined();
      expect(conversations).toStrictEqual([
        {
          createdAt: '2023-06-01T12:00:00Z',
          id: sampleConversation1.id,
          listMessages: expect.any(Function),
          metadata: {},
          name: 'Test Conversation',
          onStreamEvent: expect.any(Function),
          sendMessage: expect.any(Function),
          updatedAt: '2023-08-02T12:00:00Z',
        },
        {
          createdAt: '2024-09-02T12:00:00Z',
          id: sampleConversation2.id,
          listMessages: expect.any(Function),
          metadata: {},
          name: 'Test Conversation2',
          onStreamEvent: expect.any(Function),
          sendMessage: expect.any(Function),
          updatedAt: '2024-09-05T12:00:00Z',
        },
      ]);
      expect(nextToken).toBe(sampleNextToken);
      // #endregion assertions
    });
  });

  describe('Messages', () => {
    // #region mocking common
    const sampleConversation = {
      id: 'conversation-id',
    };
    const sampleConversationMessage1 = {
      content: [{ text: 'foo' }],
      conversationId: sampleConversation.id,
      createdAt: '2024-08-22T18:28:00.596Z',
      id: 'message-id',
      role: 'user',
    };
    const sampleConversationMessage2 = {
      content: [{ text: 'foo' }],
      conversationId: sampleConversation.id,
      createdAt: '2024-08-22T18:28:00.596Z',
      id: 'message-id',
      role: 'user',
    };
    const sampleConversationStreamEvent = {
      id: 'stream-event-id',
      conversationId: sampleConversation.id,
      associatedUserMessageId: sampleConversationMessage1.id,
      contentBlockDeltaIndex: 0,
      contentBlockDoneAtIndex: undefined,
      contentBlockIndex: 0,
      contentBlockText: 'foo',
      stopReason: undefined,
      contentBlockToolUse: undefined,
    };
    // #endregion mocking common

    test('Send a message', async () => {
      // #region mocking
      const sampleSendMessageInput = {
        content: [{ text: 'foo' }],
        aiContext: { arbitrary: 'data' },
        toolConfiguration: {
          tools: {
            myTool: {
              inputSchema: {
                json: {
                  type: 'object',
                  properties: {
                    toolUseParam: {
                      type: 'string',
                    },
                  },
                },
              },
            },
          },
        } as const,
      };
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getConversation: sampleConversation,
          },
        },
        {
          data: {
            sendMessage: sampleConversationMessage1,
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);

      const client = generateClient<Schema>();
      // get conversation
      const { data: conversation } = await client.conversations.chatBot.get({
        id: sampleConversation.id,
      });
      // send conversation message
      const { data: message, errors: sendMessageErrors } =
        (await conversation?.sendMessage(sampleSendMessageInput)) ?? {};
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(sendMessageErrors).toBeUndefined();
      expect(message).toStrictEqual(sampleConversationMessage1);
      // #endregion assertions
    });

    test('List messages', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getConversation: sampleConversation,
          },
        },
        {
          data: {
            listMessages: {
              items: [sampleConversationMessage1, sampleConversationMessage2],
            },
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);

      const client = generateClient<Schema>();
      // get conversation
      const { data: conversation } = await client.conversations.chatBot.get({
        id: sampleConversation.id,
      });
      // send conversation message
      const { data: messages, errors: listMessagesErrors } =
        (await conversation?.listMessages()) ?? {};
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(listMessagesErrors).toBeUndefined();
      expect(messages).toStrictEqual([
        sampleConversationMessage1,
        sampleConversationMessage2,
      ]);
      // #endregion assertions
    });

    test('Paginate messages', async () => {
      // #region mocking
      const sampleNextToken = 'next-token';
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            getConversation: sampleConversation,
          },
        },
        {
          data: {
            listMessages: {
              items: [sampleConversationMessage1, sampleConversationMessage2],
              nextToken: sampleNextToken,
            },
          },
        },
      ]);
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);

      const client = generateClient<Schema>();
      // get conversation
      const { data: conversation } = await client.conversations.chatBot.get({
        id: sampleConversation.id,
      });
      // send conversation message
      const {
        data: messages,
        nextToken,
        errors: listMessagesErrors,
      } = (await conversation?.listMessages({
        limit: 100,
        nextToken: 'previous-next-token',
      })) ?? {};
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(listMessagesErrors).toBeUndefined();
      expect(messages).toStrictEqual([
        sampleConversationMessage1,
        sampleConversationMessage2,
      ]);
      expect(nextToken).toBe(sampleNextToken);
      // #endregion assertions
    });

    test('Subscribe to messages', async () => {
      // #region mocking
      const { spy, subSpy, subs, generateClient } = mockedGenerateClient([
        {
          data: {
            getConversation: sampleConversation,
          },
        },
        {
          data: {
            sendMessage: sampleConversationMessage1,
          },
        },
      ]);
      const mockHandler = jest.fn();
      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking
      // #region api call
      // App.tsx
      Amplify.configure(config);

      const client = generateClient<Schema>();
      // get conversation
      const { data: conversation } = await client.conversations.chatBot.get({
        id: sampleConversation.id,
      });
      // subscribe to messages
      conversation?.onStreamEvent({
        next: (streamEvent) => {
          mockHandler(streamEvent);
        },
        error: () => {},
      });

      subs.onCreateAssistantResponseChatBot.next({
        data: {
          onCreateAssistantResponseChatBot: {
            ...sampleConversationStreamEvent,
          },
        },
      });

      await pause(1);
      // #endregion api call
      // #region assertions
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(subOptionsAndHeaders(subSpy)).toMatchSnapshot();
      const {
        contentBlockText,
        contentBlockToolUse,
        ...rest
      } = sampleConversationStreamEvent;
      const expectedConversationStreamEvent = {
        text: contentBlockText,
        toolUse: contentBlockToolUse,
        ...rest,
      };
      expect(mockHandler).toHaveBeenCalledWith(expectedConversationStreamEvent);
      // #endregion assertions
    });

    test('Uses custom conversation handler', () => {
      const customConversationHandlerMock = {
        eventVersion: '1.0',
        getInstance: jest.fn()
      } as const;

      const schema = a.schema({
        SampleChat: a
          .conversation({
            aiModel: a.ai.model('Claude 3 Haiku'),
            systemPrompt: 'testSystemPrompt',
            handler: customConversationHandlerMock
          }),
      });

      const transformedSchema = schema.transform().schema;
      expect(transformedSchema).toMatchSnapshot();
      const expectedDirective = '@conversation(aiModel: "anthropic.claude-3-haiku-20240307-v1:0", systemPrompt: "testSystemPrompt", handler: { functionName: "FnSampleChat", eventVersion: "1.0" })';
      expect(transformedSchema).toContain(expectedDirective);
    });
  });
});
