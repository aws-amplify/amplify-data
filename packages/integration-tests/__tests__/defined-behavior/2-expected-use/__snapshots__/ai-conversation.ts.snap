// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AI Conversation Routes Conversations Create a conversation 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateConversationChatBotInput!) { createConversationChatBot(input: $input) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "input": {
          "metadata": "{"arbitrary":"data"}",
          "name": "Test Conversation",
        },
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "1",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Conversations Get a conversation 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getConversationChatBot(id: $id) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "id": "conversation-id",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "2",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Conversations List conversations 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelConversationChatBotFilterInput,$limit: Int,$nextToken: String) { listConversationChatBots(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id name metadata createdAt updatedAt owner } nextToken __typename } }",
      "variables": {},
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "3",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Conversations Paginate conversations 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelConversationChatBotFilterInput,$limit: Int,$nextToken: String) { listConversationChatBots(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id name metadata createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "limit": 100,
        "nextToken": "previous-next-token",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "3",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Conversations Update a conversation 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdateConversationChatBotInput!) { updateConversationChatBot(input: $input) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "input": {
          "id": "conversation-id",
          "metadata": "{"arbitrary":"data"}",
          "name": "updated conversation name",
        },
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "9",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Messages List messages 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getConversationChatBot(id: $id) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "id": "conversation-id",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "2",
        "category": "ai",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelConversationMessageChatBotFilterInput,$limit: Int,$nextToken: String) { listConversationMessageChatBots(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id conversationId role content { text document { format name source { bytes } } image { format source { bytes } } toolResult { content { document { format name source { bytes } } image { format source { bytes } } json text } toolUseId status } toolUse { toolUseId name input } } aiContext toolConfiguration { tools { toolSpec { name description inputSchema { json } } } } createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "filter": {
          "conversationId": {
            "eq": "conversation-id",
          },
        },
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "6",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Messages Paginate messages 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getConversationChatBot(id: $id) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "id": "conversation-id",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "2",
        "category": "ai",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelConversationMessageChatBotFilterInput,$limit: Int,$nextToken: String) { listConversationMessageChatBots(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id conversationId role content { text document { format name source { bytes } } image { format source { bytes } } toolResult { content { document { format name source { bytes } } image { format source { bytes } } json text } toolUseId status } toolUse { toolUseId name input } } aiContext toolConfiguration { tools { toolSpec { name description inputSchema { json } } } } createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "filter": {
          "conversationId": {
            "eq": "conversation-id",
          },
        },
        "limit": 100,
        "nextToken": "previous-next-token",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "6",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Messages Send a message 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getConversationChatBot(id: $id) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "id": "conversation-id",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "2",
        "category": "ai",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    mutation($conversationId: ID!, $content: [ContentBlockInput], $aiContext: AWSJSON, $toolConfiguration: ToolConfigurationInput) {
      chatBot(conversationId: $conversationId, content: $content, aiContext: $aiContext, toolConfiguration: $toolConfiguration) {id conversationId role content { text document { format name source { bytes } } image { format source { bytes } } toolResult { content { document { format name source { bytes } } image { format source { bytes } } json text } toolUseId status } toolUse { toolUseId name input } } aiContext toolConfiguration { tools { toolSpec { name description inputSchema { json } } } } createdAt updatedAt owner}
    }
  ",
      "variables": {
        "aiContext": "{"arbitrary":"data"}",
        "content": [
          {
            "text": "foo",
          },
        ],
        "conversationId": "conversation-id",
        "toolConfiguration": {
          "tools": [
            {
              "toolSpec": {
                "description": undefined,
                "inputSchema": {
                  "json": "{"type":"object","properties":{"toolUseParam":{"type":"string"}}}",
                },
                "name": "myTool",
              },
            },
          ],
        },
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "5",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Messages Subscribe to messages 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getConversationChatBot(id: $id) { id name metadata createdAt updatedAt owner } }",
      "variables": {
        "id": "conversation-id",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "2",
        "category": "ai",
      },
    },
    {},
  ],
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    subscription($conversationId: ID!) {
      onCreateAssistantResponseChatBot(conversationId: $conversationId) {id conversationId role content { text document { format name source { bytes } } image { format source { bytes } } toolResult { content { document { format name source { bytes } } image { format source { bytes } } json text } toolUseId status } toolUse { toolUseId name input } } aiContext toolConfiguration { tools { toolSpec { name description inputSchema { json } } } } createdAt updatedAt owner}
    }
  ",
      "variables": {
        "conversationId": "conversation-id",
      },
      Symbol(INTERNAL_USER_AGENT_OVERRIDE): {
        "action": "7",
        "category": "ai",
      },
    },
    {},
  ],
]
`;

exports[`AI Conversation Routes Messages Subscribe to messages 2`] = `
[
  [
    {
      "authMode": undefined,
      "query": "subscription ($conversationId: ID!) {
  onCreateAssistantResponseChatBot(conversationId: $conversationId) {
    id
    conversationId
    role
    content {
      text
      document {
        format
        name
        source {
          bytes
        }
      }
      image {
        format
        source {
          bytes
        }
      }
      toolResult {
        content {
          document {
            format
            name
            source {
              bytes
            }
          }
          image {
            format
            source {
              bytes
            }
          }
          json
          text
        }
        toolUseId
        status
      }
      toolUse {
        toolUseId
        name
        input
      }
    }
    aiContext
    toolConfiguration {
      tools {
        toolSpec {
          name
          description
          inputSchema {
            json
          }
        }
      }
    }
    createdAt
    updatedAt
    owner
  }
}
",
      "variables": {
        "conversationId": "conversation-id",
      },
    },
    {},
  ],
]
`;
